core/src/storage2/collections/stash/mod.rs

```diff
@@ -466,6 +466,26 @@ where
         }
     }
 
+    /// Removes the element stored at the given index if any.
+    ///
+    /// This method acts similar to the take API and even still returns an Option.
+    /// However, it guarantees to make no contract storage reads to the indexed
+    /// element and will only write to its internal low-level lazy cache that the
+    /// element at the given index is going to be removed at the end of the contract
+    /// execution.
+    ///
+    /// Calling this method with an index out of bounds for the returns `None` and
+    /// does not `remove` the element, otherwise it returns `Some(())`.
+    pub fn remove(&mut self, at: Index) -> Option<()> {
+        if at >= self.len_entries() {
+            // Early return since `at` index is out of bounds.
+            return None
+        }
+        self.entries.put(at, None);
+        self.header.len -= 1;
```

Comment:
Hmmm this is actually problematic since we don't really know if decreasing this counter is correct. It might be incorrect if we removed an already removed (vacant) entry for example. So this is actually pretty bad and I haven't thought about this before. It might even be that we cannot implement this functionality for `storage::Stash` because of this. Also if we remove an occupied entry we have to relink the vacant entries since they form something like a doubly-linked list of vacant entries in order to quickly find a position for insertion upon `put`. So yeah, this is a bit more involved than I initially thought it'd be.